### PR TITLE
Add controller Service Instance test

### PR DESCRIPTION
**Pull request by jasiu001**
Related to [#3113](https://github.com/kyma-project/kyma/issues/3113). Controller tests for ServiceInstance.

### CODE REVIEW

File: pkg/controller/controller_flow_instance_test.go [hunk 1]
```diff
@@ -140,3 +143,764 @@ func TestServiceInstanceDeleteWithAsyncUpdateInProgress(t *testing.T) {
...
+
+// TestCreateServiceInstanceNonExistentClusterServiceBroker tests creating a
+// ServiceInstance whose broker does not exist.
+func TestCreateServiceInstanceNonExistentClusterServiceBroker(t *testing.T) {
+	// GIVEN
+	ct := newControllerTest(t)
+	defer ct.TearDown()
+
+	require.NoError(t, ct.CreateClusterServiceClass())
+	require.NoError(t, ct.WaitForClusterServiceClass())
```
Comment (piotrmiskiewicz) [inline pkg/controller/controller_flow_instance_test.go:177]: do we need to wait for the class? the same for plan

File: pkg/controller/controller_flow_instance_test.go [hunk 2]
```diff
@@ -140,3 +143,764 @@ func TestServiceInstanceDeleteWithAsyncUpdateInProgress(t *testing.T) {
...
+			require.NoError(t, ct.WaitForInstanceCondition(condition))
+
+			// Assert that the latest generation has been observed
+			// compared to original test this `ObservedGeneration` should be equal to 0
+			// because in original test apiserver is used and its functionality set `instance.Generation` to 1
+			// inside file `pkg/registry/servicecatalog/instance/strategy.go` method `PrepareForCreate()`
+			// in this test the `instance.Generation` is not update so in `pkg/controller/controller_instance.go`
+			// in method `reconcileServiceInstanceAdd` will not increase `ObservedGeneration` value
+			// because `instance.Status.ObservedGeneration != instance.Generation` condition is not met
+			require.NoError(t, ct.EnsureInstanceObservedGeneration(0))
```
Comment (piotrmiskiewicz) [inline pkg/controller/controller_flow_instance_test.go]: does not make sense

File: pkg/controller/controller_flow_instance_test.go [hunk 3]
```diff
@@ -140,3 +143,764 @@ func TestServiceInstanceDeleteWithAsyncUpdateInProgress(t *testing.T) {
...
+			require.NoError(t, ct.CreateServiceInstance())
+
+			// THEN
+			condition := v1beta1.ServiceInstanceCondition{
+				Type:   v1beta1.ServiceInstanceConditionReady,
+				Status: v1beta1.ConditionTrue,
+				Reason: "ProvisionedSuccessfully",
+			}
+			require.NoError(t, ct.WaitForInstanceCondition(condition))
+			require.NoError(t, ct.CheckFeatureGate(state.enableFeatureGate))
```
Comment (piotrmiskiewicz) [inline pkg/controller/controller_flow_instance_test.go]: Where do you set the feature gate?
the name CheckFeatureGate is saying - check the feature gate, but I'm looking deeper and there is a logic with dashboardURL - let's discuss structure the test

File: pkg/controller/case_test.go [hunk 4]
```diff
@@ -855,6 +867,687 @@ func (ct *controllerTest) v1Now() *metav1.Time {
...
+	defer ct.fakeOSBClient.Unlock()
+
+	ct.fakeOSBClient.DeprovisionReaction = fakeosb.DynamicDeprovisionReaction(
+		func(_ *osb.DeprovisionRequest) (*osb.DeprovisionResponse, error) {
+			return &osb.DeprovisionResponse{}, nil
+		})
+}
+
+// EnsureBrokerActionExist makes sure specific fake osb client action exist
+func (ct *controllerTest) EnsureBrokerActionExist(actionType fakeosb.ActionType) error {
```
Comment (piotrmiskiewicz) [inline pkg/controller/case_test.go]: there are methods like: NumberOfOSBProvisionCalls which can be used

File: pkg/controller/case_test.go [hunk 5]
```diff
@@ -855,6 +867,687 @@ func (ct *controllerTest) v1Now() *metav1.Time {
...
+			return fmt.Errorf("expected broker action %q exist", actionType)
+		}
+	}
+
+	return nil
+}
+
+// EnsureLastUpdateBrokerActionHasCorrectPlan makes sure osb client action with type "UpdateInstance"
+// contains specific plan ID in request body parameters
+func (ct *controllerTest) EnsureLastUpdateBrokerActionHasCorrectPlan(planID string) error {
```
Comment (piotrmiskiewicz) [inline pkg/controller/case_test.go]: create a method which returns last update planID or create a method something like AssertLastOSBUpdatePlanID

File: pkg/controller/controller_flow_instance_test.go [hunk 6]
```diff
@@ -140,3 +143,764 @@ func TestServiceInstanceDeleteWithAsyncUpdateInProgress(t *testing.T) {
...
+			generation, err := ct.UpdateCustomServiceInstanceParameters(
+				state.updateParams,
+				state.updateParamsFromSecret,
+				state.deleteParams,
+				state.deleteParamsFromSecret)
+			require.NoError(t, err)
+			require.NoError(t, ct.WaitForServiceInstanceProcessedGeneration(generation))
+
+			// THEN
+			require.NoError(t, ct.EnsureBrokerActionExist(fakeosb.UpdateInstance))
```
Comment (piotrmiskiewicz) [inline pkg/controller/controller_flow_instance_test.go]: change ensure to NumberOf...

File: pkg/controller/controller_flow_instance_test.go [hunk 7]
```diff
@@ -140,3 +143,764 @@ func TestServiceInstanceDeleteWithAsyncUpdateInProgress(t *testing.T) {
...
+
+			// If the provision failed with a terminating failure
+			if state.failReason != "" {
+				condition = v1beta1.ServiceInstanceCondition{
+					Type:   v1beta1.ServiceInstanceConditionFailed,
+					Status: v1beta1.ConditionTrue,
+					Reason: state.failReason,
+				}
+				require.NoError(t, ct.EnsureServiceInstanceHasCondition(condition))
+				require.NoError(t, ct.EnsureBrokerActionNotExist(fakeosb.DeprovisionInstance))
```
Comment (piotrmiskiewicz) [inline pkg/controller/controller_flow_instance_test.go]: require.Zero(t, ct.NumberOfOSBDeprovisionCalls())

File: pkg/controller/controller_flow_instance_test.go [hunk 8]
```diff
@@ -140,3 +143,764 @@ func TestServiceInstanceDeleteWithAsyncUpdateInProgress(t *testing.T) {
...
+			if state.orphanMitigation {
+				if state.orphanMitigationReason == "" {
+					state.orphanMitigationReason = "ProvisionCallFailed"
+				}
+				condition = v1beta1.ServiceInstanceCondition{
+					Type:   v1beta1.ServiceInstanceConditionOrphanMitigation,
+					Status: v1beta1.ConditionTrue,
+					Reason: state.orphanMitigationReason,
+				}
+				require.NoError(t, ct.EnsureServiceInstanceHasCondition(condition))
```
Comment (piotrmiskiewicz) [inline pkg/controller/controller_flow_instance_test.go]: use WaitForInstanceCondition

File: pkg/controller/controller_flow_instance_test.go [hunk 9]
```diff
@@ -140,3 +143,764 @@ func TestServiceInstanceDeleteWithAsyncUpdateInProgress(t *testing.T) {
...
+
+// TestCreateServiceInstanceFailsWithNonexistentPlan tests creating a ServiceInstance whose ServicePlan
+// does not exist
+func TestCreateServiceInstanceFailsWithNonexistentPlan(t *testing.T) {
+	// GIVEN
+	ct := newControllerTest(t)
+	defer ct.TearDown()
+
+	ct.SetupEmptyPlanListForOSBClient()
+	require.NoError(t, ct.CreateServiceBrokerWithIncreaseRelistRequests())
```
Comment (piotrmiskiewicz) [inline pkg/controller/controller_flow_instance_test.go]: ct.CreateSimpleClusterServiceBroker(),

why do you need increasing relist requests

File: pkg/controller/controller_flow_instance_test.go [hunk 10]
```diff
@@ -140,3 +143,764 @@ func TestServiceInstanceDeleteWithAsyncUpdateInProgress(t *testing.T) {
...
+			if state.useExternalNames {
+				generation, err = ct.UpdateServiceInstanceExternalPlanName(testOtherPlanExternalID)
+			} else {
+				generation, err = ct.UpdateServiceInstanceInternalPlanName(testOtherClusterServicePlanName)
+			}
+			require.NoError(t, err)
+			require.NoError(t, ct.WaitForServiceInstanceProcessedGeneration(generation))
+
+			// THEN
+			require.NoError(t, ct.EnsureBrokerActionExist(fakeosb.UpdateInstance))
```
Comment (piotrmiskiewicz) [inline pkg/controller/controller_flow_instance_test.go]: ct.NumberOfUpdateCalls (you need to create such method, see other NumberOf... methods)

File: pkg/controller/controller_flow_instance_test.go [hunk 11]
```diff
@@ -140,3 +143,764 @@ func TestServiceInstanceDeleteWithAsyncUpdateInProgress(t *testing.T) {
...
+				state.updateParams,
+				state.updateParamsFromSecret,
+				state.deleteParams,
+				state.deleteParamsFromSecret)
+			require.NoError(t, err)
+			require.NoError(t, ct.WaitForServiceInstanceProcessedGeneration(generation))
+
+			// THEN
+			require.NoError(t, ct.EnsureBrokerActionExist(fakeosb.UpdateInstance))
+			require.NoError(t, ct.EnsureBrokerActionWithParametersExist(state.expectedParams))
```
Comment (piotrmiskiewicz) [inline pkg/controller/controller_flow_instance_test.go]: ct.AssertLastUpdateAction, see AssertLastBindRequest

File: pkg/controller/controller_flow_instance_test.go [hunk 12]
```diff
@@ -140,3 +143,764 @@ func TestServiceInstanceDeleteWithAsyncUpdateInProgress(t *testing.T) {
...
+				if state.orphanMitigationReason == "" {
+					state.orphanMitigationReason = "ProvisionCallFailed"
+				}
+				condition = v1beta1.ServiceInstanceCondition{
+					Type:   v1beta1.ServiceInstanceConditionOrphanMitigation,
+					Status: v1beta1.ConditionTrue,
+					Reason: state.orphanMitigationReason,
+				}
+				require.NoError(t, ct.EnsureServiceInstanceHasCondition(condition))
+				require.NoError(t, ct.EnsureServiceInstanceOrphanMitigationStatus(true))
```
Comment (piotrmiskiewicz) [inline pkg/controller/controller_flow_instance_test.go]: AssertServiceinstanceOrphanMitigationStatus

File: pkg/controller/controller_flow_instance_test.go [hunk 13]
```diff
@@ -140,3 +143,764 @@ func TestServiceInstanceDeleteWithAsyncUpdateInProgress(t *testing.T) {
...
+			// Assert that the observed generation is up-to-date, that orphan mitigation is not in progress,
+			// and that the instance is not in a failed state.
+			require.NoError(t, ct.EnsureObservedGenerationIsCorrect())
+			require.NoError(t, ct.EnsureServiceInstanceOrphanMitigationStatus(false))
+			condition = v1beta1.ServiceInstanceCondition{
+				Type:   v1beta1.ServiceInstanceConditionFailed,
+				Status: v1beta1.ConditionFalse,
+			}
+			require.NoError(t, ct.EnsureServiceInstanceHasNoCondition(condition))
+			require.NoError(t, ct.EnsureBrokerActionExist(fakeosb.ProvisionInstance))
```
Comment (piotrmiskiewicz) [inline pkg/controller/controller_flow_instance_test.go]: use ct.NumberOfProvisionCalls

**Review by piotrmiskiewicz — COMMENTED** [review]
Create methods like AssertXXX(t *testing.T, ....) instead of Ensure

File: pkg/controller/controller_flow_instance_test.go (hunk 1 above)
Comment (jasiu001) [reply pkg/controller/controller_flow_instance_test.go:177]: The original test waits for [class](https://github.com/kubernetes-incubator/service-catalog/blob/master/test/integration/controller_instance_test.go#L188) and [plans](https://github.com/kubernetes-incubator/service-catalog/blob/master/test/integration/controller_instance_test.go#L209). In my opinion they wanted to be sure provision failed due lack of ServiceBroker not class or plans. Should it be omitted? I am not sure

File: pkg/controller/case_test.go [hunk 14]
```diff
@@ -855,6 +869,617 @@ func (ct *controllerTest) v1Now() *metav1.Time {
...
+			plans = append(plans, *request.PlanID)
+		}
+	}
+
+	return plans
+}
+
+// AssertBrokerActionWithParametersExist makes sure osb client action with type "UpdateInstance"
+// contains specific parameters in request body parameters
+func (ct *controllerTest) AssertBrokerActionWithParametersExist(parameters map[string]interface{}) error {
```
Comment (piotrmiskiewicz) [inline pkg/controller/case_test.go]: it asserts update action - fix the name

File: pkg/controller/case_test.go [hunk 15]
```diff
@@ -855,6 +869,617 @@ func (ct *controllerTest) v1Now() *metav1.Time {
...
+
+	ct.fakeOSBClient.DeprovisionReaction = fakeosb.DynamicDeprovisionReaction(
+		func(_ *osb.DeprovisionRequest) (*osb.DeprovisionResponse, error) {
+			return &osb.DeprovisionResponse{}, nil
+		})
+}
+
+// AssertLastOSBUpdatePlanID makes sure osb client action with type "UpdateInstance"
+// contains specific plan ID in request body parameters
+func (ct *controllerTest) AssertLastOSBUpdatePlanID() error {
```
Comment (piotrmiskiewicz) [inline pkg/controller/case_test.go]: please, refactor all Assert* methods to have one argument t *testing.T, see existing AssertOSBBasicAuth

File: pkg/controller/controller_flow_instance_test.go [hunk 16]
```diff
@@ -140,3 +142,762 @@ func TestServiceInstanceDeleteWithAsyncUpdateInProgress(t *testing.T) {
...
+				Status: v1beta1.ConditionTrue,
+				Reason: "ProvisionedSuccessfully",
+			}
+			require.NoError(t, ct.WaitForInstanceCondition(condition))
+
+			if state.enableFeatureGate {
+				require.NoError(t, ct.AssertServiceInstanceDashboardURL())
+			} else {
+				require.NoError(t, ct.AssertServiceInstanceEmptyDashboardURL())
+			}
```
Comment (piotrmiskiewicz) [inline pkg/controller/controller_flow_instance_test.go:742]: try such test:
```
t.Run(tn, func(t *testing.T) {
			// GIVEN
			ct := newControllerTest(t)
			defer ct.TearDown()

			//if state.enableFeatureGate {
			oldValue := utilfeature.DefaultFeatureGate.Enabled(scfeatures.UpdateDashboardURL)
			utilfeature.DefaultFeatureGate.Set(fmt.Sprintf("%v=%v", scfeatures.UpdateDashboardURL, state.enableFeatureGate))
			defer utilfeature.DefaultFeatureGate.Set(fmt.Sprintf("%v=%v", scfeatures.UpdateDashboardURL, oldValue))
			//}

			require.NoError(t, ct.CreateSimpleClusterServiceBroker())
			require.NoError(t, ct.WaitForReadyBroker())

			// WHEN
			require.NoError(t, ct.CreateServiceInstance())


			// THEN
			require.NoError(t, ct.WaitForReadyInstance())

			ct.SetUpdateServiceInstanceResponseWithDashboardURL()

			ct.UpdateServiceInstanceParameters()

			time.Sleep(time.Second)

			if state.enableFeatureGate {
				require.NoError(t, ct.AssertServiceInstanceDashboardURL())
			} else {
				require.NoError(t, ct.AssertServiceInstanceEmptyDashboardURL())
			}
		})
```

**Review by piotrmiskiewicz — CHANGES_REQUESTED** [review]

**Comment by jasiu001** [comment]
PR moved to main fork [#2667](https://github.com/kubernetes-sigs/service-catalog/pull/2667)

**Pull request closed without merging**